ENH: `escape` html argument in Styler.format

attack68 opened the pull request:
- [x] closes #40388 

Adds an HTML `escape` parameter to `Styler.format`, which acts to escape the data value **before** it is passed to a formatter.

For example, `Styler.format('<a href="a.com/{0}">{0}</a>', escape=True)` will not escape the HTML as part of the formatting function, it will only escape the value `{0}` going into the formatter.

attack68 committed:
update escaping for html

attack68 committed:
update escaping for html

attack68 committed:
tests for format_escape

jreback commented on pandas/io/formats/style.py:
```diff
@@ -113,6 +114,10 @@ class Styler:
+    escape : bool, default False
+        Replace the characters ``&``, ``<``, ``>``, ``'``, and ``"`` in cell display
+        strings with HTML-safe sequences.
+
```
versionadded 1.3

jreback reviewed (changes requested):
sure, minor comment. ping on green.

jreback commented:
cc @jnothman if comments.

jreback commented:
also checks are failing

jnothman reviewed (approved):
LGTM

attack68 committed:
Merge remote-tracking branch 'upstream/master' into escape_html_formatter

attack68 committed:
check and docs fix

attack68 commented:
> also checks are failing

when fixing these I noticed that the `na_rep` was not working with `escape` since missing values were escaped before being replaced.

so I refactored to fix this and added a test for it.

@jreback ping is green.

attack68 committed:
versionadded and whatsnew

attack68 commented on pandas/io/formats/style.py:
```diff
@@ -2171,9 +2206,12 @@ def _maybe_wrap_formatter(
     else:
-        return lambda x: na_rep if pd.isna(x) else formatter_func(x)
+        if escape:
+            return lambda x: na_rep if pd.isna(x) else formatter_func(escape_func(x))
```
This might be a bug escaping floats, since escaping converts to string. Will review and come back

attack68 replied:
OK this is fixed now too. 3rd test added.

attack68 committed:
only use escape on str

attack68 committed:
refactor

attack68 commented on pandas/io/formats/style.py:
```diff
@@ -2170,10 +2205,17 @@ def _maybe_wrap_formatter(
+        """only use escape_func on str, else return input"""
+        return escape_func(x) if isinstance(x, str) else x
+
     if na_rep is None:
```
hows this?

jreback replied:
good

attack68 commented:
this should be good to go in now, I believe.

jreback commented:
thanks @attack68
